Automattic/jetpack
# Repeat Visitor Block: Add JS and PHP fixture tests

Related to: 237-gh-Automattic/view-design and https://github.com/Automattic/jetpack/issues/19812

This follows on from the PHP fixture tests added for the Business Hours block in: https://github.com/Automattic/jetpack/pull/19743 and tracked in the primary issue for adding PHP fixture tests for server rendered Jetpack blocks: https://github.com/Automattic/jetpack/issues/19812

The approach in this PR proposes adding a parent test case class with a method that we can hopefully share across multiple server-rendered blocks for generating and testing the server-rendered block test fixtures.

#### Changes proposed in this Pull Request:

For the Repeat Visitor block:

* Add JS-based fixture tests
* Add tests for the Edit UI of the block
* Add PHP-based fixture tests that include testing both states of the block's server rendering (below the threshold to reveal the block, and above the threshold to reveal the block)

#### Jetpack product discussion


237-gh-Automattic/view-design and https://github.com/Automattic/jetpack/issues/19812

#### Does this pull request change what data or activity we track or use?


No

#### Testing instructions:






These are similar instructions as to the ones included in #19743 

* Ensure the PHP tests are passing in CI
* Ensure the JS tests are passing in CI

##### Testing locally

1. Check out this change locally, and following the instructions in this repo, set up your local Docker development environment and run the following from the root directory of this repo:

```
yarn docker:phpunit --testsuite=extensions
```

2. Delete the generated server-rendered fixtures included in this PR, and run the above command again. You should see that the test throws an error, and reports back the fixtures that have been generated:

![image](https://user-images.githubusercontent.com/14988353/120145873-8ec94400-c227-11eb-82fc-e5d87514ea44.png)

3. Run the command again, and the tests should pass
4. Manually make a change to one of the server-rendered fixtures files and re-run the testing command. It should fail, complaining that the strings don't match. But, it should also tell you which fixture it failed on with a (hopefully) friendlier message than just the strings not matching:

![image](https://user-images.githubusercontent.com/14988353/120146019-c9cb7780-c227-11eb-95b7-9f6824815476.png)

5. Check that the JS tests are passing:

From the root of the Jetpack repo:

```
cd projects/plugins/jetpack
yarn run jest projects/plugins/jetpack/extensions/blocks/repeat-visitor/test/
```

6. Finally, while this PR shouldn't change the block at all, give it a quick smoke test to make sure the one tiny addition (exporting the `RepeatVisitorEdit` component) hasn't broken anything.
